macrotypes/examples/tests/mlish-tests.rkt

```diff
@@ -359,13 +359,16 @@
 ;; nested lambdas
 
 (check-type (λ ([x : X]) (λ ([y : X]) y)) : (→/test X (→ X X)))
-(check-not-type (λ ([x : X]) (λ ([y : X]) y)) : (→/test {X} X (→/test {Y} Y Y)))
-(check-type (λ ([x : X]) (λ ([y : Y]) y)) : (→/test {X} X (→/test {Y} Y Y)))
+(check-not-type (λ ([x : X]) (λ ([y : X]) y)) : (→/test X (→ Y Y)))
+(check-type (λ ([x : X]) (λ ([y : Y]) y)) : (→/test X (→ Y Y)))
 (check-not-type (λ ([x : X]) (λ ([y : Y]) x)) : (→/test X (→ X X)))
 
 (check-type 
   ((λ ([x : X]) (λ ([y : Y]) y)) 1)
-  : (→/test Y Y))
```

Comment:
Would the original test still pass?

Reply:
No, that's one of the remaining issues.

With @AlexKnauth's change which [lifts nested ∀ from the function's return type](https://github.com/stchang/macrotypes/pull/29/commits/dbafd4e9556b6e3799af893757482886ace124d6), the outer lambda has type `(∀ (X Y) (→ X (→ Y Y)))`, but then the value restriction kicks in, and prevents Y from appearing in covariant positions (in case the body was a let-over-lambda capturing the Y type).

If I revert that commit, then the outer lambda's type is `(∀ (X) (→ X (∀ (Y) (→ Y Y))))` (which sounds good to me as I used Typed Racket a lot, but @AlexKnauth I'm not sure if it is okay to revert that commit). The problem then becomes that `(∀ (Y) (→ Y Y))` cannot be annotated as `(→ Int Int)`, because the former is not a subtype of the former (an explicit instantiation is needed instead).

So I'm trying to see if I can find some combination of changes to `solve` + `add-constraints` + subtyping which makes everything work.

Reply:
Ok, yes I'm playing around a bit more and see the behavior you mention. I'm starting to remember now and I think that was one of the original holdups because it means programs like `(((λ ([x : X]) (λ ([y : Y]) y)) 1) 1)` will be rejected which is not ideal.

Thanks for the update and thanks for work on this!

Reply:
The type of the outer lambda should be `(∀ (X Y) (→ X (→ Y Y)))` so that type inference can instantiate either `X` or `Y` if it needs to. Here it needs to instantiate `X` as `Int`, but `Y` isn't constrained. So why is `Y` also solved to `Int` here?

Reply:
@AlexKnauth I'm not sure what you mean by "why is Y also solved to Int here?", could you clarify? Thanks!

Tentative answer: In that test, the `(check-type … : (→ Int Int))` places an expected-type on the outer lambda's return value, which allows the `Y` to be instantiated to `Int` (but then the return type cannot be left as `(→ Y Y)`, as the value restriction forces it to be monomorphic, IIUC).

Reply:
Oh, right. So the test could just as easily be `(check-type … : (→ Int Int))` and it would still pass. The `Y` in the answer we want should a non-generalizable type variable, like the ones notated `'_a` in the "Relaxing the Value Restriction" paper.